### DISCUSSION TITLE
Automation support

### DISCUSSION THREAD

**Discussion by g-v-egidy** [memtest86plus/memtest86plus#138] [Enhancement]
I opened this issue to discuss how best to design an interface to automate calling memtest and reporting back the results. Once the design is agreed on, the code to implement it could be developed by me or my colleagues and sent as pull request.

### Automation use cases

Some use cases I want to use memtest automation for:

- Testing new PCs during production or burn-in test
- Regularly test the memory of workstations during off-hours

### Automation sequence

- memtest will be started by a boot manager, for example grub or EFI shell
- the boot manager will save the information that memtest was booted (for example with `load_env` / `save_env` in grub)
- an option to use the automation interface will be given to memtest on the boot command line, for example `autorun=1`
- memtest runs the number of full test runs given on the commandline
- in case of sucessful test: reboot
- in case an error was found: stop and show the errors found. The operator will notice the machine not
  automatically rebooting after the alotted time and see the message on the screen
- the boot manager will see that memtest was started before (`load_env`) and now boot a dedicated Linux OS
- the Linux system will read out the report data (see below) and BIOS event log and store or transmit it for further processing

### Result transmission

The automation system should work for UEFI and systems with classic BIOS. 

While UEFI would offer some convenient interfaces to store data on the EFI system partition, the classic BIOS doesn't offer anything comparable. Implementing storage and file system access for systems with classic BIOS is non trivial. Since the automation sequence shown above would always boot into a adapted Linux system afterwards, we can change a commonly available system parameter in memtest and read it out and reset it in the Linux system.

A commonly available system parameter to transmit the result information with could be the system date offered by the realtime clock (RTC). Every PC provides this, the interface to it is common among all PCs and the data usually contained in it can be restored by the Linux system with a call to ntpdate and hwclock.

So when memtest starts, it sets the date for example to January 2010. Should a hardware problem or memory error result in a spurious reset, the Linux will detect the year 2010 and report an error. When memtest successfully passed all tests, it sets for example January 2011. If this date is seen by the Linux system, success is reported.

### Discussion

We are using the system date as result transmission variable with a patched version of memtest86+ v5 for several years in our production environment now and it works reliably.

Do you think result transmission via date is a good idea and getting something based on this upstream is viable?

Or do you have a better idea for transmitting test results?

**Comment by x86fr** [comment]
Automation is probably the most requested features for "pro" users (while Secure boot support is the top priority for mainstream users)

As you said, the main issue is to pass a non-volatile information from Memtest86+ to the outside world. Many of the betatesters are working in big factories, with tens of thousands of computers assembled each months. The basic process they use is the "autoreboot on success". If a pass fail, Memtest86+ will just stop forever, waiting for a user input.

I've worked on an automation process using PXE. The first time the tftp server sees a new MAC address, it send the Memtest86+ binary. If it sees the same MAC address requesting a boot binary more than 15 minutes after the initial request, it deduces that everything went well and send the OS installation binary instead. If something went wrong, Memtest86+ should be waiting for a user action with a FAIL banner displayed. It's far from ideal, but it works 99% of the time. 

Other than that, the easiest and (almost) universal way to pass information to the outside world is to use the serial COM device. But even if a COM header is often present even on the latest motherboards, it involves additional manipulation.

The ideal solution would be to use the USB mass-storage specification to write a report on FAT formatted drive. This unfortunately requires some serious coding skill, even if Martin has already cleared the path with USB controller supports. 

If working on USB mass-storage is too complex, an alternative temporary solution would be to find a NVRAM somewhere. As you said, messing with the basic CMOS (and especially the date/time) is a quick & dirty solution that might work (this could help: https://github.com/JackieXie168/nvram). I have nothing about quick & dirty solutions (I love them most of the time), but you can encounter two critical issues here : the first one is the legacy method used to alter the cmos data. On modern computer (even for date/time that are not covered by a checksum), it can be detected as a CMOS corruption and led to a CMOS clear (try on a computer without CSM that is only intended to have these fields changed via UEFI runtime services). The second one is Secure Boot. I'm not sure that such a trick will not be considered as a security issue and prevent the code being signed. 

That's why I strongly suggest finding another NVRAM as there is probably many of them on a modern computer that can be used as a scratchpad. Some are real NVRAM (ie: DMI entries), others are pseudo-NVRAM only able to retain information on a warm reset (ie: the value of a stopped counter, chipset scratchpad registers, ...).

PS: Unrelated (or maybe...) but Memtest86+ can probably read and write easily to a mass storage device *before* UEFI boot services are exited. I plan to use this to read a configuration file at init time, but it won't work as soon as the test is started.

**Reply by 01e3** [reply]
Would leaving the report simply in RAM and using something similar to [kexec / kdump](https://www.kernel.org/doc/html/latest/admin-guide/kdump/kdump.html) would be an option?

If we are allowed to carry over something like 10-20MBs for the kernel+initrd, which Memtest86+ would take care of not destroying during testing, we could have a complete OS available with proper mass storage support (including filesystems like NTFS/FAT/exFAT) and even (especially) the network. Once Memtest86+ completes, based on the provided criteria, it can boot the OS which would save the report (PASS/FAIL) or upload it via the network.

**Reply by g-v-egidy** [reply]
> I've worked on an automation process using PXE. The first time the tftp server sees a new MAC address, it send the Memtest86+ binary. If it sees the same MAC address requesting a boot binary more than 15 minutes after the initial request, it deduces that everything went well and send the OS installation binary instead.

There are two ways that this could miss a failure:
- a spurious reboot because of bad hardware, or a uncorrectable double memory error detected by ECC but reading out ECC status doesn't work. A reboot could also be triggered by a hardware watchdog when the system hangs.
- booting memtest didn't work, for example due to some transmission error in TFTP (you sometimes get them, seldom, but it happens when you do it often enough). The bios has some internal watchdog that checks if a OS has been booted within some time (like 15 minutes) and does a reboot if not. I have seen this implemented on some machines.

Missing these cases would not be acceptable for me.

> As you said, messing with the basic CMOS (and especially the date/time) is a quick & dirty solution that might work (this could help: https://github.com/JackieXie168/nvram). I have nothing about quick & dirty solutions (I love them most of the time), but you can encounter two critical issues here : the first one is the legacy method used to alter the cmos data. On modern computer (even for date/time that are not covered by a checksum), it can be detected as a CMOS corruption and led to a CMOS clear (try on a computer without CSM that is only intended to have these fields changed via UEFI runtime services). 

There is a difference between static nvram data and the realtime clock. Static nvram data has become much more complicated in the last years and modifying it became an involved process. It is also very dependant on the bios vendor (AMI, Insyde, coreboot,...).

But the rtc stayed simple and compatible to the 286 / AT PCs. Even today linux directly writes to the registers and does not use EFI runtime services to modify the date. See https://github.com/torvalds/linux/blob/master/drivers/rtc/rtc-cmos.c and https://github.com/torvalds/linux/blob/master/drivers/rtc/rtc-mc146818-lib.c

So this is something memtest can do too and does not need to worry about checksums or possible corruption. As I wrote we are using this approach successfully with a wide variety of systems in a production environment for years. Here is the relevant part of our patch to memtest:

```
+//If there are error sets RTC year to 2010 and if not to 2012.
+// (10)d => (00010000)bcd => 0x10
+// (12)d => (00010010)bcd => 0x12
+void set_rtc_date(bool is_error)
+{
+    int new_date;
+    if (is_error)
+        new_date = 0x10;
+    else
+        new_date = 0x12;
+    //minutes
+    outb(0x82, 0x70);
+    outb(0x00, 0x71);
+    //hours
+    outb(0x84, 0x70);
+    outb(0x00, 0x71);
+    //day
+    outb(0x87, 0x70);
+    outb(0x01, 0x71);
+    //month
+    outb(0x88, 0x70);
+    outb(0x01, 0x71);
+    //year
+    outb(0x89, 0x70);
+    outb(new_date, 0x71);
+}
```

**Reply by g-v-egidy** [reply]
> If we are allowed to carry over something like 10-20MBs for the kernel+initrd, which Memtest86+ would take care of not destroying during testing, we could have a complete OS available with proper mass storage support (including filesystems like NTFS/FAT/exFAT) and even (especially) the network. Once Memtest86+ completes, based on the provided criteria, it can boot the OS which would save the report (PASS/FAIL) or upload it via the network.

You mean you want the bootloader (like grub) append the linux kernel + initrd to memtest, just like cpu microcode updates and initrd are now appended when booting linux? Or how would the kernel+initrd get into ram so that memtest can kexec them?

How complicated is doing a kexec? Do you need to set up virtual memory addressing in a special way or something similar before you can do it?

**Reply by 01e3** [reply]
> > If we are allowed to carry over something like 10-20MBs for the kernel+initrd, which Memtest86+ would take care of not destroying during testing, we could have a complete OS available with proper mass storage support (including filesystems like NTFS/FAT/exFAT) and even (especially) the network. Once Memtest86+ completes, based on the provided criteria, it can boot the OS which would save the report (PASS/FAIL) or upload it via the network.
> 
> You mean you want the bootloader (like grub) append the linux kernel + initrd to memtest, just like cpu microcode updates and initrd are now appended when booting linux? Or how would the kernel+initrd get into ram so that memtest can kexec them?

Yes, precisely - we can use the initrd mechanism for a bootloader like grub, syslinux or iPXE to provide the additional image(s).

Alternative, they could be simply embedded  in the binary as another section(s), but this seems less flexibly and harder to maintain.

> How complicated is doing a kexec? Do you need to set up virtual memory addressing in a special way or something similar before you can do it?
Not sure TBH, we would need to look. That said, I'm also not sure if we even *need* kexec. At the end, Memtest86+ can just work as an intermediate loader by making sure the kernel image and the initrd are in the proper place in the memory, updating the pointers in the headers and for example calling the start_of_setup routine pointed by the "setup" header.

We would also need to decide which mechanism to use to pass the report - initrd? specific place in the memory? cmd_line? Note that the decision does not need to be final - the mechanism would live in the Memtest86+ binary and the initrd image, so can be changed later. As a simple start, we can begin with cmdline and if needed, work later on a something more advanced.

**Reply by x86fr** [reply]
Please note that we will never "reserve" a 10-20MB memory area that will just stay there untouched and can't be tested.
This memory area must be relocated (several times) to allow testing of all the RAM.

**Reply by x86fr** [reply]
> But the rtc stayed simple and compatible to the 286 / AT PCs. Even today linux directly writes to the registers and does not use EFI runtime services to modify the date. See https://github.com/torvalds/linux/blob/master/drivers/rtc/rtc-cmos.c and https://github.com/torvalds/linux/blob/master/drivers/rtc/rtc-mc146818-lib.c

There is several other drivers also used by the Linux Kernel to set the RTC date, based on numerous factors.
Here is the one for EFI RTC : https://github.com/torvalds/linux/blob/master/drivers/rtc/rtc-efi.c

**Reply by g-v-egidy** [reply]
> There is several other drivers also used by the Linux Kernel to set the RTC date, based on numerous factors. Here is the one for EFI RTC : https://github.com/torvalds/linux/blob/master/drivers/rtc/rtc-efi.c

The rtc-efi module is only used on non-x86-platforms:
```
config RTC_DRV_EFI
	tristate "EFI RTC"
	depends on EFI && !X86
```

I think it was initially developed for IA-64 / Itanium, but is maybe now also used on some UEFI-based aarch64 machines.

Does memtest target any other architectures than i686 and x86_64 in the near or medium term? I doesn't look to me that way, so I don't think we should base a decision on something relevant only for other architectures.

**Reply by 01e3** [reply]
> Please note that we will never "reserve" a 10-20MB memory area that will just stay there untouched and can't be tested.

Completely agree. I'm definitely not proposing to leave this untouched.

> This memory area must be relocated (several times) to allow testing of all the RAM.

That would be my assumption as well. Do you thin it would be feasible?

**Reply by g-v-egidy** [reply]
> > This memory area must be relocated (several times) to allow testing of all the RAM.
> 
> That would be my assumption as well. Do you this it would be feasible?

This looks quite invasive to me. This is not just some patch that modifies the startup and pass-loop handling of memtest, but something that touches most actual tests. My guess is that this will constantly complicate maintenance and further development of memtest.

So all in all the kexec solution looks to me quite involved:
- "take in" appended linux + initrd memory block, determine how big it is and how it is splitted between kernel and initrd
- copy the linux+initrd around during the tests so that the space they take up can still be tested
- find out how to do a kexec, implement it for i686 and x86_64
- implement creation and passing of a boot commandline to the kexec'ed linux

The solution to modify the RTC date is of course a bit of a hack. But it needs just minimal additional code in memtest and it works on i686, x86_64 and classic bios and UEFI, without any architecture/boot mode specific code.

So would you consider accepting a rtc date based automation solution, of course given that code quality of a pull request is good?

**Reply by 01e3** [reply]
By all means, RTC based solution is definitely going to be much faster and easier to implement. The question is for how long it is going to last. My concern is that this is yet another old x86 subsystem it is eventually going to be removed. But perhaps this is something we should worry about *once* it happens.

**Reply by x86fr** [reply]
> So would you consider accepting a rtc date based automation solution, of course given that code quality of a pull request is good?

Messing with the RTC to save an information sounds really too nasty to be added to main. 
We just can't alter RTC of the user PC and hope that an NTP service will set it back properly later in an invisible way. 
Many things can go wrong, from firewall blocking NTP to badly developed software lock that can trigger to unwanted cron starting unexpectedly. 

To work in real-world application, it's also a modification that need a be implemented on a wider process than just memtest86+ alone.

This said, this could be maintained in a separate branch as a temporary "experimental" solution for automation.

**Reply by g-v-egidy** [reply]
> Messing with the RTC to save an information sounds really too nasty to be added to main. We just can't alter RTC of the user PC and hope that an NTP service will set it back properly later in an invisible way.

I don't suggest to always modify the RTC, just when a special commandline option is given to memtest. Also autoreboot on pass could be a separate commandline option than the RTC-result-transmission:

- `autoreboot=<n>`: this would make memtest automatically reboot after n successful passes
- `rtc_result`: this would activate to modify the RTC based on test result. Without this option, the RTC isn't touched.

I think if the user explicitly requested this with a commandline option it is not too nasty, isn't it?

> To work in real-world application, it's also a modification that need a be implemented on a wider process than just memtest86+ alone.

You mean provide something like a set of scripts and short documentation how to set the whole thing up, including the linux that parses the results? This could be done. But I think it would be more like a template the user can take to start, but then adapt to their needs, since the details of how to do the automation differ a bit, see for example the way I described with grub on disk vs. the way of doing it via PXE you described.

The scripts provided as template could for example first test if ntp works and the rtc can be modified successfully before starting memtest in this mode and otherwise abort with an error message.

**Comment by g-v-egidy** [comment]
I created pull requests #206 and #207 that implement the autoreboot and rtc_result commandline options.

A detailed documentation is included in the rtc_result patch.
